4016-debug-extension: support debugging of extensions using Debug panel

danarad05 opened the pull request:
Signed-off-by: Dan Arad <dan.arad@sap.com>

#### What it does
fixed #4016 with this functionality:
1. Supports start, restart, stop of pwa-extensionHost launch type on debug toolbar.
2. Utilizes Hosted Plugin functionality.

#### How to test
_**Note**: can only be tested on `browser` as `electron` doesn't support Hosted Plugin functionality currently (see #2868)_

1. Add pwa-extensionHost configuration to launch configurations like so:
![image](https://user-images.githubusercontent.com/71069170/97988704-0cd68100-1de6-11eb-85ec-7dc9c456185f.png)
2. start debugging from debug toolbar
3. two threads will eventually appear after the hosted plugin windows will be opened - like so:
![image](https://user-images.githubusercontent.com/71069170/97988834-40b1a680-1de6-11eb-93cf-ee18312caeee.png)
4. you can choose restart or stop in debug toolbar.

Approved CQ for copied code: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=22801

#### Review checklist

- [X] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

danarad05 commented:
FYI: @amiramw @DoroNahari

vince-fugnitto reviewed (commented):
@danarad05 I do not believe the `@theia/debug` extension should know about plugins.

kittaakos commented:
I won't have the bandwidth to review the changes this week, but as I see, it requires a CQ. Please start the CQ process as documented [here](https://github.com/eclipse-theia/theia/wiki/Registering-CQs#codecontent-is-copied-from-the-main-vs-code-repo), by the time we have the legal approval for the merge, hopefully, we will have the review too ;)

danarad05 commented:
> @danarad05 I do not believe the `@theia/debug` extension should know about plugins.
> 
> cc @colin-grant-work @kenneth-marut-work you may be interested in the following changes which impact debugging.

@vince-fugnitto I have created a debug ContributionProvider for hosted plugin to registering to. Issue #4016 talks about invoking hosted plugin from debug toolbar (maybe thinking eventually to removing invoking hosted plugin from command box).
@amiramw

danarad05 commented:
> e time we have the legal approval for the merge, hopefully, we will have the review too ;)

Hi @kittaakos 
1. Thanks for your input.
2. I'm not able to find the intellectual property section  - tried also searching for it and the CQ but to no avail. Please advise.
3. Furthermore, I copied a couple of very small structures - nothing major IMO - if that is an issue I can change it a bit to simplify  things... it would be a shame if this process takes time for this small of a change.

Thanks
Fyi: @amiramw

kittaakos commented:
> 2\. I'm not able to find the intellectual property section  - tried also searching for it and the CQ but to no avail. Please advise.

I do not understand this 😊 You do not know how to submit a CQ, is this the question?

> 3\. copied a couple of very small structures - nothing major IMO

I am not a lawyer nor a decision-maker, so I do one thing during the review; if there is copied code, we need a CQ:

https://github.com/eclipse-theia/theia/blob/78664d352134fa9920844b4465f2b9e22f80609d/packages/debug/src/browser/debug-plugin-contribution.ts#L32

amiramw commented:
@danarad05 i'll try to help with the CQ

danarad05 commented:
> > 2. I'm not able to find the intellectual property section  - tried also searching for it and the CQ but to no avail. Please advise.
> 
> I do not understand this 😊 You do not know how to submit a CQ, is this the question?

I meant that explanation in the [link](https://github.com/eclipse-theia/theia/wiki/Registering-CQs#codecontent-is-copied-from-the-main-vs-code-repo) you provided didn't get me to the "Intellectual Property" section etc... I don't know if it works for you - I logged in and didn't see and "Intellectual Property" section and tried searching for it. Maybe you should check if explanation there is still valid.. 
![image](https://user-images.githubusercontent.com/71069170/98084791-0d722480-1e85-11eb-9f75-ea253ddc7d3d.png)

I will try to resolve it with the help of @amiramw, Thanks.

amiramw commented:
CQ for copied code: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=22801

danarad05 commented on packages/debug/src/browser/debug-plugin-contribution.ts:
```diff
@@ -0,0 +1,49 @@
+/********************************************************************************
+ * Copyright (c) 2019 SAP SE or an SAP affiliate company and others.
```
Fixed

danarad05 commented on packages/debug/src/browser/debug-session.tsx:
```diff
@@ -73,8 +75,15 @@ export class DebugSession implements CompositeTreeElement {
         this.connection.onRequest('runInTerminal', (request: DebugProtocol.RunInTerminalRequest) => this.runInTerminal(request));
+        this.connection.onRequest('launchVSCode', (request: LaunchVSCodeRequest) => this.launchVSCode(request));
+
+        if (this.debugPluginCP) {
```
**_why do you need to members that contain the same list_**
I thought that getContributions was not saving list in it - my mistake. **Fixed**

**_why concat if we are in the constructor_**
It was at first as postConstruct so I guess I left it that way. **Fixed**

danarad05 commented on packages/debug/src/browser/debug-session.tsx:
```diff
@@ -253,13 +269,10 @@ export class DebugSession implements CompositeTreeElement {
-            } else {
-                await this.sendRequest('launch', this.configuration);
-            }
+            await this.sendRequest((this.configuration.request === 'attach' ? 'attach' : 'launch'), this.configuration);
```
I shortened the code here - if request is different from `attach ` it goes to `launch ` - not sure if this case exists or not but didn't want to change logic. Should I?

amiramw replied:
i think it can only be these two
![image](https://user-images.githubusercontent.com/9420921/100395527-fb317380-3049-11eb-91e6-cd1f7c612821.png)

danarad05 replied:
Fixed

danarad05 commented on packages/debug/src/browser/debug-session.tsx:
```diff
@@ -298,10 +312,14 @@ export class DebugSession implements CompositeTreeElement {
             if (!await this.exited(1000)) {
                 await this.disconnect(restart);
             }
+        } else if (this.configuration.type === 'pwa-extensionHost') {
```
Not sure what you mean. Could you elaborate?
What happens is that we enter twice this function when stopping hosted plugin - once for the "Launch Extension" thread and second time for the "Hosted Plugin" thread  - so first time it gets into 'pwa-extensionHost' part and second to last `else`.
What I couldn't find another solution to was the `this.fireExited()` line - without it it would not complete exiting the "Launch Extension" thread and not sure why (`'exited'` was not fired)

amiramw replied:
why do you enter this function twice?

danarad05 commented on packages/debug/src/browser/debug-session.tsx:
```diff
@@ -375,6 +405,22 @@ export class DebugSession implements CompositeTreeElement {
+    protected async launchVSCode({ arguments: { args } }: LaunchVSCodeRequest): Promise<LaunchVSCodeResult> {
+        let result = {};
+        for (const contrib of this.debugPluginContribs) {
+            if (contrib.debug) {
```
Fixed

danarad05 commented on packages/debug/src/browser/debug-session.tsx:
```diff
@@ -375,6 +405,22 @@ export class DebugSession implements CompositeTreeElement {
+                const instanceURI = await contrib.debug(this.getDebugPluginConfig(args));
+                if (instanceURI) {
+                    const instanceURL = new URL(instanceURI);
+                    if (instanceURL.port) {
```
port is not really being used currently - I just mimicked the structures/result from vscode to get similar behavior.

danarad05 commented on packages/debug/src/browser/debug-session.tsx:
```diff
@@ -73,8 +75,15 @@ export class DebugSession implements CompositeTreeElement {
         protected readonly messages: MessageClient,
-        protected readonly fileService: FileService) {
+        protected readonly fileService: FileService,
+        protected readonly debugPluginCP: ContributionProvider<DebugPluginContribution>
```
@amiramw 
could you give an example for what you mean?

amiramw replied:
DebugPluginContribution is very specific for debug plugin flow. also it doesn't make sense that it is a list as there will always be one. it would be better if contribution is generic and plugin debug is only a specific implementation of it.

danarad05 replied:
done.

amiramw commented:
> CQ for copied code: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=22801

If I understand correctly the CQ is approved.

danarad05 commented:
@vince-fugnitto @akosyakov 
cc: @amiramw 
Hi,
A question I hope you can help me with (or if you can direct me to someone who might help that would be also great):
I'm starting debugging a `pwa-extensionHost` config type whicheventually creates 4 threads created under Threads pane like so:
![image](https://user-images.githubusercontent.com/71069170/101500280-ed97ba00-3976-11eb-801c-174b17eb266b.png)

When I try to stop debugging - one thread is disposed of but not the others. If I'm clicking stop again - then it disposes all the others. 

From what I can see last thread received 'terminated' event (because clicking on stop debug button selects by default last thread to stop) and so it was disposed successfully - BUT previous thread (thread #3 in list) has not received 'terminated' event by DebugSession and therefore it isn't disposed of.
If on the other hand I select first thread and then click 'stop' - all threads are disposed of. 

So my questions are:
1. You have any idea/input concerning why 'terminated' event is not received by DebugSession for 3rd thread in list?
2. Is it a viable solution in your eyes that when clicking 'stop' debug - first thread will be selected by default instead of last one?

Any suggestions are welcomed. I tried different approaches but to no avail.

Thanks,
-Dan

amiramw commented:
@danarad05 i don't think you need to address this behavior in this PR as it is not unique for hosted plugin debug.

danarad05 commented:
@vince-fugnitto @amiramw ready for your review. 
p.s. I didn't solve issues with debug stop not closing all threads.

danarad05 commented:
@vince-fugnitto @amiramw 
https://github.com/eclipse-theia/theia/pull/8706#issuecomment-742494199
reminder

danarad05 commented:
@vince-fugnitto @amiramw
#8706 (comment)
reminder

vince-fugnitto commented on packages/debug/src/browser/debug-session-contribution.ts:
```diff
@@ -114,6 +113,8 @@ export class DefaultDebugSessionFactory implements DebugSessionFactory {
     @inject(FileService)
     protected readonly fileService: FileService;
+    @inject(ContributionProvider) @named(DebugContribution)
+    protected readonly debugCP: ContributionProvider<DebugContribution>;
```
I believe we should be clear in the naming of the variables:

```suggestion
    protected readonly debugContribution: ContributionProvider<DebugContribution>;
```

The same is true for all other occurrences of `debugCP` across the pull-request.

danarad05 replied:
What would you like it to be named? it should be `debugContributionProvider` but it seems long to me

danarad05 replied:
renamed as `debugContributionProvider`

vince-fugnitto commented on packages/debug/src/browser/debug-session-manager.ts:
```diff
@@ -273,10 +274,13 @@ export class DebugSessionManager {
+
     protected async doRestart(session: DebugSession, restart?: any): Promise<DebugSession | undefined> {
         if (await session.restart()) {
             return session;
```
Changes in the file can be reverted.

danarad05 replied:
done

vince-fugnitto commented on packages/debug/src/browser/debug-session.tsx:
```diff
@@ -73,8 +73,12 @@ export class DebugSession implements CompositeTreeElement {
         protected readonly messages: MessageClient,
-        protected readonly fileService: FileService) {
+        protected readonly fileService: FileService,
+        protected readonly debugCP: ContributionProvider<DebugContribution>
```
This is a breaking change which is not documented in the `changelog`.
Downstream clients who use `DebugSession` or extend it will be broken by the `constructor`.

marechal-p replied:
TBF it doesn't look like this is a public API (meant to be overriden without breakage). Still nice to add to the CHANGELOG.

danarad05 replied:
> I tested the feature using `pwa-extensionHost` and it worked correctly but I'm wondering why the standard `extensionHost` **type** is not supported yet (which is the default when generating an extension):
> 
> ```json
> {
>     "name": "Run Extension",
>     "type": "extensionHost",
>     "request": "launch",
>     "args": [
>         "--extensionDevelopmentPath=${workspaceFolder}"
>     ],
>     "outFiles": [
>         "${workspaceFolder}/out/**/*.js"
>     ],
>     "preLaunchTask": "${defaultBuildTask}"
> },
> ```

I don't know if and when extensionHost **type** worked. Currently **pwa-extensionHost** in `vscode-js-debug` builtin plugin replaced it. See:
- https://github.com/microsoft/vscode-js-debug/blob/master/OPTIONS.md
- https://github.com/microsoft/vscode-js-debug/blob/4747389c7d50639738aeb33c1e2f6e2980d2bd0f/COMMON_PROBLEMS.md#my-app-doesnt-run-using-a-node-10-release-before-around-10180

danarad05 replied:
added changelog update without breaking changes

vince-fugnitto reviewed (commented):
I tested the feature using `pwa-extensionHost` and it worked correctly but I'm wondering why the standard `extensionHost` **type** is not supported yet (which is the default when generating an extension):

```json
{
    "name": "Run Extension",
    "type": "extensionHost",
    "request": "launch",
    "args": [
        "--extensionDevelopmentPath=${workspaceFolder}"
    ],
    "outFiles": [
        "${workspaceFolder}/out/**/*.js"
    ],
    "preLaunchTask": "${defaultBuildTask}"
},
```

marechal-p commented on packages/plugin-dev/src/browser/hosted-plugin-controller.ts:
```diff
@@ -247,7 +247,7 @@ export class HostedPluginController implements FrontendApplicationContribution {
 
     private getErrorMessage(error: Error): string {
-        return error.message.substring(error.message.indexOf(':') + 1);
+        return error?.message?.substring(error.message.indexOf(':') + 1);
```
This function returns `string` but what you wrote seem to indicate that it could return `undefined`. Something looks off with this return line.

danarad05 replied:
@marechal-p Thanks, you are correct - my bad about returning undefined. I added this because of in certain cases error or error.message would be undefined which resulted in exception which would cause debug failure where you needed to reload browser to solve it. So the check is IMO essential - I'm just adding || '' to the end like so:
```
private getErrorMessage(error: Error): string {
        return error?.message?.substring(error.message.indexOf(':') + 1) || '';
    }
```

marechal-p commented on packages/debug/src/browser/debug-session.tsx:
```diff
@@ -73,8 +73,12 @@ export class DebugSession implements CompositeTreeElement {
+        protected readonly debugCP: ContributionProvider<DebugContribution>
+    ) {
         this.connection.onRequest('runInTerminal', (request: DebugProtocol.RunInTerminalRequest) => this.runInTerminal(request));
+        this.registerDebugContributions(options.configuration.type, this.connection);
```
So every time a new `DebugSession` is created, all contributions will be registered again? I would have expected registration to be a one time thing?

danarad05 replied:
There are 2 different parts happening here:
1. "Registration" of contributions - as you called it - are contribution providers/containers populated when app loads - once - see **DebugMainImpl** - packages\plugin-ext\src\main\browser\debug\debug-main.ts.

2. **registerDebugContributions** method here is going over list of contributions and invoking `register` method on contribution itself in order for contribution itself to register to requests/events sent from `vscode-js-debug`.  This method is not populating contributions again.

danarad05 commented:
> ```
> I tested the feature using `pwa-extensionHost` and it worked correctly but I'm wondering why the standard `extensionHost` **type** is not supported yet (which is the default when generating an extension):
> 
> ```json
> {
>     "name": "Run Extension",
>     "type": "extensionHost",
>     "request": "launch",
>     "args": [
>         "--extensionDevelopmentPath=${workspaceFolder}"
>     ],
>     "outFiles": [
>         "${workspaceFolder}/out/**/*.js"
>     ],
>     "preLaunchTask": "${defaultBuildTask}"
> },
> ```

I don't know if and when extensionHost **type** worked. Currently **pwa-extensionHost** in `vscode-js-debug` builtin plugin replaced it. See:
- https://github.com/microsoft/vscode-js-debug/blob/master/OPTIONS.md
- https://github.com/microsoft/vscode-js-debug/blob/4747389c7d50639738aeb33c1e2f6e2980d2bd0f/COMMON_PROBLEMS.md#my-app-doesnt-run-using-a-node-10-release-before-around-10180

vince-fugnitto commented:
> I don't know if and when extensionHost **type** worked. Currently **pwa-extensionHost** in `vscode-js-debug` builtin plugin replaced it. See:
> 
> * https://github.com/microsoft/vscode-js-debug/blob/master/OPTIONS.md
> * https://github.com/microsoft/vscode-js-debug/blob/4747389c7d50639738aeb33c1e2f6e2980d2bd0f/COMMON_PROBLEMS.md#my-app-doesnt-run-using-a-node-10-release-before-around-10180

@danarad05 I believe that `extensionHost` should be supported, not only `pwa-extensionHost` which only limits debugging a small subset of extensions:

https://github.com/microsoft/vscode/blob/e2bce32da4b7345e776e1b54caf40e30e7da4ad7/src/vs/workbench/contrib/debug/common/debugUtils.ts#L61-L63

Moreover, the `extensionHost` is the default when generating extensions (using `yo code` for example), and is the extension type of all the `vscode-builtins` from which the framework consumes:

- https://github.com/microsoft/vscode/blob/master/extensions/css-language-features/.vscode/launch.json#L6
- https://github.com/microsoft/vscode/blob/master/extensions/html-language-features/.vscode/launch.json#L12
- https://github.com/microsoft/vscode/blob/aaf5a7fee3b577a421c2ad0b90b565b331b793b5/extensions/emmet/.vscode/launch.json#L8

The following was just a subset of extension for example purposes. Since the framework supports the debug configurations defined in `.vscode/launch.json`, I would not expect end-users to prefix `pwa-` before their type when debugging with the framework. 

Please let me know whether you believe this is outside the scope of the pull-request, but we are limiting what the framework can successfully debug, and it looks like `extensionHost` is the more common `type`.

danarad05 commented:
> @danarad05 I believe that `extensionHost` should be supported, not only `pwa-extensionHost` which only limits debugging a small subset of extensions:

@vince-fugnitto 

Debugged using `extensionHost` and saw that `PluginDebugAdapterContribution` is executing `resolveDebugConfiguration` which replaces `extensionHost` with `pwa-extensionHost` and continues debug session with this type.

Did you try debugging with `extensionHost` and it didn't work for you?
 
**original type = extensionHost**

![image](https://user-images.githubusercontent.com/71069170/105969408-e5175200-6090-11eb-985f-1669797e1120.png)

**resolved type: pwa-extensionHost**

![image](https://user-images.githubusercontent.com/71069170/105969933-78e91e00-6091-11eb-89ea-d2a1ac890229.png)

vince-fugnitto commented:
> > @danarad05 I believe that `extensionHost` should be supported, not only `pwa-extensionHost` which only limits debugging a small subset of extensions:
> 
> @vince-fugnitto
> 
> Debugged using `extensionHost` and saw that `PluginDebugAdapterContribution` is executing `resolveDebugConfiguration` which replaces `extensionHost` with `pwa-extensionHost` and continues debug session with this type.
> 
> Did you try debugging with `extensionHost` and it didn't work for you?

@danarad05 correct, it unfortunately did not work for me until I prefixed it by `pwa-` which is not the case in vscode.

My launch configuration looks like the following (to test I used the `hello-world` example generated from `yo code`):

```json
{
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Run Extension",
      "type": "extensionHost",
      "request": "launch",
      "args": [
        "--extensionDevelopmentPath=${workspaceFolder}"
      ],
      "outFiles": [
        "${workspaceFolder}/out/**/*.js"
      ]
    }
  ]
}
```

1. start the application with the generated hello-world extension
2. attempt to start the hosted instance with the `run extension` debug configuration
3. the debug session starts but does not spawn the hosted extension (in a separate window)
4. updated the config to use `pwa-extensionHost` instead of `extensionHost`
5. the debug configuration executes successfully, the hosted plugin is launched in a separate window

danarad05 commented:
> @danarad05 correct, it unfortunately did not work for me until I prefixed it by `pwa-` which is not the case in vscode.

@vince-fugnitto you're a magician Vince :).. can't seem to reproduce problem - you are on Linux?

vince-fugnitto commented:
> > @danarad05 correct, it unfortunately did not work for me until I prefixed it by `pwa-` which is not the case in vscode.
> 
> @vince-fugnitto you're a magician Vince :).. can't seem to reproduce problem - you are on Linux?

Interestingly I retried the changes on both linux and macos (both electron and browser applications), and it seemed to work with both `pwa-extensionHost` and `extensionHost` as a `type` in the debug launch configuration. :+1: I'm not sure why I had issues initially both times I verified the pull-request (unless something was fixed).

DukeNgn reviewed (commented):
@danarad05 I'm testing on Windows and the problem @vince-fugnitto mentioned is there. But it happens with or without `pwa-` for me. 

1. Start Theia, open workspace in this hello-world folder 
[hello-world-vscode-ext.zip](https://github.com/eclipse-theia/theia/files/5887865/hello-world-vscode-ext.zip)
2. `run extension` debug configuration from the debug view.
3.  The debug session starts but doesn't spawn anything.

After a while, it times out.

```
root INFO [hosted-plugin: 12672] Debug configuration provider has been registered: pwa-chrome
root INFO [hosted-plugin: 12672] Debug configuration provider has been registered: pwa-msedge
Debug configuration provider has been registered: pwa-extensionHost
Debug configuration provider has been registered: pwa-node
Debug configuration provider has been registered: node-terminal
root INFO [hosted-plugin: 12672] Debug configuration provider has been registered: pwa-chrome
Debug configuration provider has been registered: pwa-msedge
Debug configuration provider has been registered: pwa-node
Debug configuration provider has been registered: pwa-node
root ERROR Uncaught Exception:  Error: Hosted plugin instance is not running.
root ERROR Error: Hosted plugin instance is not running.
    at ElectronNodeHostedPluginRunner.AbstractHostedInstanceManager.terminate (C:\Users\ducngn\Documents\workspaces\testPR\theia\packages\plugin-dev\lib\node\hosted-instance-manager.js:196:19)
    at Timeout._onTimeout (C:\Users\ducngn\Documents\workspaces\testPR\theia\packages\plugin-dev\lib\node\hosted-instance-manager.js:366:27)
    at listOnTimeout (internal/timers.js:531:17)
    at processTimers (internal/timers.js:475:7)

```

**Note**: I am able to start in vs code with this extension

danarad05 commented:
> After a while, it times out.

@DukeNgn 
Thanks for reviewing. Unfortunately I'm having difficulty reproducing issue. Took you zip and ran it successfully, also re-cloned branch to different location and that succeeded also. 
Trying to understand what could be difference between.

danarad05 commented:
> The problem can be reproduced with Electron target. Browser version works for me.

@DukeNgn 
cc: @amiramw 
Seems to me that (at least on windows) **Electron** doesn't support debugging hosted plugins currently. Problem is reproduced with no connection to this PR. 

Tried to debug and see if can find a quick solution but didn't find one which fully works - found same solution (shell = true for windows) as suggested in #2868 [comment](https://github.com/eclipse-theia/theia/issues/2868#issuecomment-691376991) - which opened another electron window but didn't work completely.

I would suggest opening a separate bug for this issue or joining it with #2868

danarad05 commented:
@vince-fugnitto @kittaakos @marechal-p @amiramw @DukeNgn 

Guys, it's been 3 months + since this PR was opened. Could we advance this PR?
Thanks

vince-fugnitto reviewed (commented):
@danarad05 I have some bandwidth to perform a review, please do the following:
- resolve the conflicts and formatting errors of the `changelog`
- please provide more information in the commit message (I believe I've mentioned this previously) so we can accurately describe the changes and how they work
- please updated the pull-request description to state that the changes should be tested solely on the `browser` as I believe you've mentioned that the changes do not work correctly on electron yet due to issues currently on master

danarad05 commented on packages/debug/src/browser/debug-session-manager.ts:
```diff
@@ -201,6 +201,7 @@ export class DebugSessionManager {
     }
```
Fixed

danarad05 commented:
> @danarad05 I have some bandwidth to perform a review, please do the following:
> 
> * resolve the conflicts and formatting errors of the `changelog`
> * please provide more information in the commit message (I believe I've mentioned this previously) so we can accurately describe the changes and how they work
> * please updated the pull-request description to state that the changes should be tested solely on the `browser` as I believe you've mentioned that the changes do not work correctly on electron yet due to issues currently on master

@vince-fugnitto 
Done. Thanks

vince-fugnitto commented on CHANGELOG.md:
```diff
@@ -31,6 +31,7 @@
 - [scm] updated code required to highlight nodes on search in the `ScmTreeWidget` [#8929](https://github.com/eclipse-theia/theia/pull/8929)
 - [task] fixed issue where tasks were not successfully executed without `cwd` explicitly set [#8949](https://github.com/eclipse-theia/theia/pull/8949)
 - [terminal] reduced the severity of certain terminal logs [#8908](https://github.com/eclipse-theia/theia/pull/8908)
+- [debug] added support for managing debug sessions for extensions from debug panel (previously only possible using `Hosted Plugin` commands) [#8706](https://github.com/eclipse-theia/theia/pull/8706)
```
The following entry should be under `1.11.0` since `1.10.0` has already been released.

danarad05 replied:
@vince-fugnitto 
Fixed. Thanks

vince-fugnitto reviewed (commented):
@danarad05 can you confirm that the following use-case should/should not be handled by this pull-request:

1. start the debug session - it spawns the hosted window
2. close the hosted window
3. the debug session is not terminated, the `stop` action does not terminate the session

_Screenshare_:


https://user-images.githubusercontent.com/40359487/108722009-a7231780-74f0-11eb-86bf-4ad527ca90d0.mp4

danarad05 commented:
> @danarad05 can you confirm that the following use-case should/should not be handled by this pull-request:
> 
> 1. start the debug session - it spawns the hosted window
> 2. close the hosted window
> 3. the debug session is not terminated, the `stop` action does not terminate the session
> 
> _Screenshare_:

@vince-fugnitto 
Please see my previous comment on the matter - https://github.com/eclipse-theia/theia/pull/8706#issuecomment-740679918

vince-fugnitto commented:
> @vince-fugnitto
> Please see my previous comment on the matter - [#8706 (comment)](https://github.com/eclipse-theia/theia/pull/8706#issuecomment-740679918)

Can you then update the pull-request, the following is untrue?
> Supports start, restart, stop of pwa-extensionHost launch on debug toolbar.

Stopping the debug session is not supported (I am unable to do so when testing). Should we not aim for at least minimal functionality, at the moment only `starting` debug session is supported. 

In addition, restarting I end up in this state:

![image](https://user-images.githubusercontent.com/40359487/108726348-4f3adf80-74f5-11eb-8fe9-515aa6a92859.png)

danarad05 commented:
> > @vince-fugnitto
> > Please see my previous comment on the matter - [#8706 (comment)](https://github.com/eclipse-theia/theia/pull/8706#issuecomment-740679918)
> 
> Can you then update the pull-request, the following is untrue?
> 
> > Supports start, restart, stop of pwa-extensionHost launch on debug toolbar.
> 
> Stopping the debug session is not supported (I am unable to do so when testing). Should we not aim for at least minimal functionality, at the moment only `starting` debug session is supported.

1. It's not totally accurate - `Stop` is supported in same way that `Hosted Plugin: Stop instance` is supported as it is invoking same logic.
2. I I suggested in  [#8706 (comment)](https://github.com/eclipse-theia/theia/pull/8706#issuecomment-740679918) -  if I stop first thread instead of `current` thread (which is never first one) then stop succeeds in closing all threads (from what I remember).
.

vince-fugnitto commented:
> t's not totally accurate - `Stop` is supported in same way that `Hosted Plugin: Stop instance` is supported as it is invoking same logic.

@danarad05 I believe it is inacurrate to say that `stop` is supported as part of this pull-request.

> Supports start, restart, stop of pwa-extensionHost launch on debug toolbar.

It is mentioned that `start`, `restart` and `stop` work through the debug toolbar but this is not the case.

---

If the goal of the pull-request is only to support `start` please document it as such. In addition, please file issues to support features not covered by this pull-request:
- stop
- restart
- step-in
- step-out
- pause

For end-users, the only functionality that works correctly with the changes is `start` unless I am doing something incorrect.

danarad05 commented:
> > t's not totally accurate - `Stop` is supported in same way that `Hosted Plugin: Stop instance` is supported as it is invoking same logic.
> 
> @danarad05 I believe it is inacurrate to say that `stop` is supported as part of this pull-request.
> 
> > Supports start, restart, stop of pwa-extensionHost launch on debug toolbar.
> 
> It is mentioned that `start`, `restart` and `stop` work through the debug toolbar but this is not the case.
> 
> If the goal of the pull-request is only to support `start` please document it as such. In addition, please file issues to support features not covered by this pull-request:
> 
> * stop
> * restart
> * step-in
> * step-out
> * pause
> 
> For end-users, the only functionality that works correctly with the changes is `start` unless I am doing something incorrect.

Again, I do not exactly agree. If you consider that `Hosted Plugin` commands are "supported" by current version of Theia (regardless of if they function totally correctly) then this PR is invoking those commands but from the debug panel. clicking on the debug panel icon/buttons invokes the same code that is invoked by the actions in the Hosted Plugin command bar and bottom bar. 
I could change phrasing and mention that start/stop/restart invoke the Hosted Plugin commands and current logic implemented by them. I have to mention that start/stop/restart buttons are part of this PR.. from your comment a user would think that just start button works which is not the case. 
Furthermore, as I suggested before - I could implement that if you click on "stop" button in debug panel (in the case that you don't select a thread prior to that) - that I will kill first thread by default and that should eliminate all other threads as well. Concerning restart - I will check.

vince-fugnitto commented:
> Again, I do not exactly agree. If you consider that `Hosted Plugin` commands are "supported" by current version of Theia (regardless of if they function totally correctly) then this PR is invoking those commands but from the debug panel. clicking on the debug panel icon/buttons invokes the same code that is invoked by the actions in the Hosted Plugin command bar and bottom bar.

Except that `hosted plugin` works at the moment if we want to start, stop and restart, but using those same commands from the debug session does not work, hence I do not believe that it is supported. I do not have the same issues by doing it through the normal `hosted plugin` workflow as I do through the debug toolbar:

https://user-images.githubusercontent.com/40359487/108852359-8374d580-75b3-11eb-83a7-2e307fdd3969.mp4

For the case of the `hosted plugin` mode, the session actually terminates for end-users.

> ...from your comment a user would think that just start button works which is not the case.

I was unable to `stop` or `restart` successfully through the debug toolbar, am I missing something?
The only way to recover the application after starting a session is completely restarting the application (we are left in a debug session state).

> Furthermore, as I suggested before - I could implement that if you click on "stop" button in debug panel (in the case that you don't select a thread prior to that) - that I will kill first thread by default and that should eliminate all other threads as well. Concerning restart - I will check.

This does not fix the underlying issue, and sounds more like a workaround.

danarad05 commented:
@vince-fugnitto 
cc: @amiramw 

You were right Vince, it wasn't working well, I remembered it working better but some time has passed since this PR was created.

Anyway, I reviewed solution again and updated it so that `stop` and `restart` work also including: 

1. In **debug-session.tsx** I fixed `disconnect` method it wouldn't invoke `fireExited` when `sendRequest` method wasn't resolving - which resulted in unterminated sessions in `Threads` panel.
2. In **debug-frontend-application-contribution.ts** I updated `currentSession` for `stop` and `restart` commands so that `terminate()` and `restart()` methods would be invoked on first session - which helped complete actions successfully.

I tested:
1. stopping (by button or menu command and without selecting a session from "Threads" list): closes all sessions successfully.
2. restarting (by button or menu command and without selecting a session from "Threads" list):  closes all debug sessions (other than first one that is already connected to node-debug2 extension) and starts new ones instead. 
3. I also tested debugging of extension after first start, and after second and third restart - all worked successfully.
4. selecting a session from `Threads` list and stopping - works - as it invokes different method than (1) - which I didn't change.

Please review again
Thanks

danarad05 commented:
@vince-fugnitto 
bump https://github.com/eclipse-theia/theia/pull/8706#issuecomment-785076241

vince-fugnitto commented on packages/debug/src/browser/debug-frontend-application-contribution.ts:
```diff
@@ -611,11 +611,17 @@ export class DebugFrontendApplicationContribution extends AbstractViewContributi
         registry.registerCommand(DebugCommands.STOP, {
-            execute: () => this.manager.currentSession && this.manager.currentSession.terminate(),
+            execute: () => {
+                this.manager.updateCurrentSession(undefined);
+                return this.manager.currentSession && this.manager.currentSession.terminate();
```
Should it not be the responsibility of the manager to update the current session, rather than callers?
I do not know if outside callers should be able to simply clear the session.

danarad05 replied:
changed it so manager will update current session.

vince-fugnitto commented on packages/plugin-dev/src/browser/hosted-plugin-controller.ts:
```diff
@@ -247,7 +247,7 @@ export class HostedPluginController implements FrontendApplicationContribution {
 
     private getErrorMessage(error: Error): string {
-        return error.message.substring(error.message.indexOf(':') + 1);
+        return error?.message?.substring(error.message.indexOf(':') + 1) || '';
```
`error` is currently non optional so why do we do:

```ts
error?.message?.substring(...)
```

danarad05 replied:
This change was required after debugging exceptions thrown in some cases which didn't contain error or error.message which would result in hiding original stack trace of error.

marechal-p replied:
Change the type of the `error` parameter then, `any` could be enough. Right now it feels weird since `error` is supposed to always be defined, as well as `error.message`. But from what you are saying this is not always the case, so we should update the typings to reflect that.

danarad05 replied:
@marechal-p Fixed. Thanks

vince-fugnitto commented on packages/plugin-dev/src/browser/hosted-plugin-manager-client.ts:
```diff
@@ -306,8 +323,52 @@ export class HostedPluginManagerClient {
+    }
+
     protected getErrorMessage(error: Error): string {
-        return error.message.substring(error.message.indexOf(':') + 1);
+        return error?.message?.substring(error.message.indexOf(':') + 1) || '';
+    }
```
Same as the previous comment.

danarad05 replied:
https://github.com/eclipse-theia/theia/pull/8706#discussion_r590097344

danarad05 replied:
Fixed

vince-fugnitto commented on CHANGELOG.md:
```diff
@@ -2,2020 +2,2027 @@
 
```
Please revert all unnecessary changes to the `changelog`.

danarad05 replied:
Fixed.

danarad05 commented:
@vince-fugnitto Please review again. Thanks

danarad05 commented:
@vince-fugnitto @marechal-p Please review latest changes. Thanks

vince-fugnitto commented on packages/plugin-ext/src/hosted/node/hosted-plugin-process.ts:
```diff
@@ -124,9 +124,9 @@ export class HostedPluginProcess implements ServerPluginRunner {
-            childProcesses.forEach(childProcess =>
-                this.killProcess(parseInt(childProcess.PID))
-            );
+            childProcesses.forEach(childProcess => {
+                this.killProcess(parseInt(childProcess.PID));
+            });
```
The change is unnecessary, it can be a single line.

danarad05 replied:
Fixed

vince-fugnitto commented on packages/debug/src/browser/debug-session-manager.ts:
```diff
@@ -287,6 +287,16 @@ export class DebugSessionManager {
         return this.start(options);
     }
 
+    async terminateSessions(): Promise<void> {
+        this.updateCurrentSession(undefined);
+        this.currentSession?.terminate();
+    }
+
+    async restartSessions(): Promise<void> {
+        this.updateCurrentSession(undefined);
+        this.currentSession?.restart();
+    }
```
I'm not sure I understand these changes. The `updateCurrentSession` sets the `currentSession` as `undefined`.
This means that `terminate` and `restart` will never occur.

danarad05 replied:
```
protected updateCurrentSession(session: DebugSession | undefined): void {
        this.currentSession = session || this.sessions[0];
    }
```
passing `undefined` updates `currentSession` to session[0] - which is what I wanted in order to terminate and restart successfully sessions.

vince-fugnitto replied:
Thank you, I had missed the part where if `undefined` it would pick up the next session.

vince-fugnitto commented on packages/debug/src/browser/debug-session-manager.ts:
```diff
@@ -287,6 +287,16 @@ export class DebugSessionManager {
         return this.start(options);
     }
 
+    async terminateSessions(): Promise<void> {
+        this.updateCurrentSession(undefined);
+        this.currentSession?.terminate();
+    }
+
+    async restartSessions(): Promise<void> {
+        this.updateCurrentSession(undefined);
+        this.currentSession?.restart();
+    }
```
The naming of the methods are misleading, we aren't dealing with multiple sessions.

danarad05 replied:
By those before me that defined debug sessions logic - each item in `Threads` section is considered a Session in Theia - therefore I named it as such. How would you like me to name them?

vince-fugnitto replied:
@danarad05 in the case we have multiple ongoing debug sessions, does this logic terminate them all (even ones which are unrelated)?

danarad05 replied:
@vince-fugnitto 
1. didn't think about that possibility. Is that a common usage - running multiple debug "launches"? if that is needed we might want to open issue for separating those n session groups from each others - in UI - as currently you cannot distinguish between them - they are all listed as equal.
2. I tested it out a little - terminating on "stop" icon (without selection a session) - terminates first session and it's child sessions. So if there are 2 different session "launches" - it terminates sessions of first group leaving second group intact. Still, there are caveats  - for instance you cannot run 2 pwa-extensionHost launches as currently extension is run on localhost:3030 (hardcoded 3030). so I wouldn't say it is full proof but probably good enough for this PR currently.

vince-fugnitto replied:
@danarad05 one example I know of in our own repo is starting two configurations as the following:
- start browser backend
- start browser frontend

The debug-view has two sessions in this instance (note that vscode does not have the `threads` view):

![image](https://user-images.githubusercontent.com/40359487/111339659-52d20a00-864e-11eb-870a-49d8f2f1f517.png)

danarad05 replied:
ok. 
First: that I guess is another issue - that Theia doesn't support "compound" launch configs like vscode.

Second: tried running both launch BE and FE which opened localhost:3000 in a new browser and worked (for the most part - as it was pretty slow and stuck a little using it). I also stop/restarted and again - it did it but I not perfectly (new `session` was added to end of list.

first run of both sessions:
![image](https://user-images.githubusercontent.com/71069170/111441125-b112f180-870f-11eb-8db6-f14167c68b85.png)

After restarting first session - a new session was added last:
![image](https://user-images.githubusercontent.com/71069170/111441311-e15a9000-870f-11eb-93a2-0dc8dad60899.png)

Third: breakpoints unfortunately in parent localhost:3000 weren't hit when running child localhost:3000.

So, I wouldn't say it's all perfect but maybe it's time to merge this PR and move forward with it.

vince-fugnitto replied:
> So, I wouldn't say it's all perfect but maybe it's time to merge this PR and move forward with it.

@danarad05 I'm just trying to understand potential implications to the debug side of the changeset and how `termination` of a debug session is now handled. Looking at VS Code's sources, they handle stopping a session like so:


```ts
async stopSession(session: IDebugSession | undefined): Promise<any> {
	if (session) {
		return session.terminate();
	}

	const sessions = this.model.getSessions();
	if (sessions.length === 0) {
		this.taskRunner.cancel();
		// User might have cancelled starting of a debug session, and in some cases the quick pick is left open
		await this.quickInputService.cancel();
		this.endInitializingState();
		this.cancelTokens(undefined);
	}

	return Promise.all(sessions.map(s => s.terminate()));
}
```

It seems in this instance, the `stop` toolbar command terminates all sessions?

- [stopHandler](https://github.com/microsoft/vscode/blob/b335d9eaf100990547ae45f3fdac54ae129dcbc3/src/vs/workbench/contrib/debug/browser/debugCommands.ts#L281-L298)
- [stopSession](https://github.com/microsoft/vscode/blob/b335d9eaf100990547ae45f3fdac54ae129dcbc3/src/vs/workbench/contrib/debug/browser/debugService.ts#L739-L754)

danarad05 replied:
@vince-fugnitto 
Seems like:
1. In VS Code there is no debug toolbar with 'play', 'stop', 'restart' commands like in Theia and commands are on "sessions" themselves:
![image](https://user-images.githubusercontent.com/71069170/111479383-981f3600-8739-11eb-96c8-5d65b43924fa.png)

2. I see also that `stop` command exist on parent sessions only (first one created upon running debug).  
![image](https://user-images.githubusercontent.com/71069170/111479832-0663f880-873a-11eb-9a97-04015adc96ed.png)

3. looking at the code you noted, stopping invokes stop on parent session - which is similar to what I did (going for first session (= sessions[0]) when user doesn't select session from Threads section) while in Theia we didn't implement "parentSession" connection between sessions.
```
// Stop should be sent to the root parent session
	while (!showSubSessions && session && session.parentSession) {
		session = session.parentSession;
	}

	await debugService.stopSession(session);
```

vince-fugnitto replied:
> 1. In VS Code there is no debug toolbar with 'play', 'stop', 'restart' commands like in Theia and commands are on "sessions" themselves:

That's not entirely true, I do see the toolbar items (I use the `docked` location - `"debug.toolBarLocation": "docked"`):

![image](https://user-images.githubusercontent.com/40359487/111481615-845bdb80-8709-11eb-96e3-27d2def9ac47.png)

> 3. looking at the code you noted, stopping invokes stop on parent session - which is similar to what I did (going for first session (= sessions[0]) when user doesn't select session from Threads section) while in Theia we didn't implement "parentSession" connection between sessions.

For my own understanding, is `sessions[0]` guaranteed to be the parent session?

danarad05 replied:
> That's not entirely true, I do see the toolbar items (I use the docked location - "debug.toolBarLocation": "docked"):

You are right. my bad, it was hidden... 

> For my own understanding, is sessions[0] guaranteed to be the parent session?

Yes, sessions[0] is parent session for first launch as it is first session created.

vince-fugnitto commented:
@amiramw do you have feedback for this pull-request as well?

amiramw commented on CHANGELOG.md:
```diff
@@ -7,7 +7,8 @@
 - [property-view] added initial version of a selection-based property-view [#8655](https://github.com/eclipse-theia/theia/pull/8655)
-  - A default implementation is available for file selections (via file navigator and default editors).
+- A default implementation is available for file selections (via file navigator and default editors).
+- [debug] added support for managing debug sessions for extensions from debug panel (previously only possible using `Hosted Plugin` commands) [#8706](https://github.com/eclipse-theia/theia/pull/8706)
```
this probably should move to 1.13.0 now

danarad05 replied:
Obviously will rebase if we are approving this PR.

danarad05 replied:
Fixed

amiramw commented on packages/debug/src/browser/debug-session.tsx:
```diff
@@ -302,22 +311,27 @@ export class DebugSession implements CompositeTreeElement {
-        if (!await this.exited(timeout)) {
-            this.fireExited(new Error(`timeout after ${timeout} ms`));
-        }
+        const TIMEOUT_MS = 1000;
```
why was it increased from 500?

danarad05 replied:
Previously we ran await this.sendRequest('disconnect', { restart }); and after it completed we ran if (!await this.exited(timeout)) { - so in reality we were waiting longer then 500ms to check if exited was received.

Now we are invoking this.sendRequest and setTimeout together - therefore I thought it would be more prudent to add a little more to timeout. I don't want that sendRequest will succeed but timeout will be thrown before it - better in my mind to let it some more rope. Also, I tested it with 500/1000ms and I think it worked better with 1000.

amiramw commented on packages/debug/src/browser/debug-session.tsx:
```diff
@@ -302,22 +311,27 @@ export class DebugSession implements CompositeTreeElement {
+        Promise.race([
+            this.sendRequest('disconnect', { restart }),
+            new Promise(reject => setTimeout(reject, TIMEOUT_MS, new Error('TIMEOUT_ERR')))
+        ]).then(res => {
```
you can use `await` and `catch` instead of `.then().catch()`

danarad05 replied:
I tried it and other options but found that setTimout throwing exception in a promise is not triggering catch but returning result - which is an error.
Please see https://javascript.info/task/error-async

amiramw commented on packages/plugin-dev/src/browser/hosted-plugin-manager-client.ts:
```diff
@@ -306,8 +323,53 @@ export class HostedPluginManagerClient {
+        const sessions = this.debugSessionManager.sessions.filter(session => session.id !== this.connection.sessionId);
+
+        if (sessions.length > 0) {
+            sessions.forEach(session => session.terminate(false));
```
add some comments about why you terminate existing sessions

danarad05 replied:
Added

amiramw commented on packages/plugin-dev/src/browser/hosted-plugin-manager-client.ts:
```diff
@@ -306,8 +323,53 @@ export class HostedPluginManagerClient {
+    private setDebugConfig(config?: DebugPluginConfiguration): DebugPluginConfiguration {
+        config = Object.assign(config || {}, { debugMode: this.hostedPluginPreferences['hosted-plugin.debugMode'] });
+        if (config.pluginLocation) {
+            this.pluginLocation = new URI((!config.pluginLocation.startsWith('/') ? '/' : '') + config.pluginLocation.replace(/\\/g, '/')).withScheme('file');
```
does plugin location have '/' also on windows?

amiramw commented on packages/plugin-dev/src/node/hosted-instance-manager.ts:
```diff
@@ -374,17 +359,14 @@ export class NodeHostedPluginRunner extends AbstractHostedInstanceManager {
-            }
+            port = process.env.HOSTED_PLUGIN_PORT ?
+                Number(process.env.HOSTED_PLUGIN_PORT) :
+                (debugConfig && debugConfig.debugPort ? Number(debugConfig.debugPort) : DEFAULT_HOSTED_PLUGIN_PORT);
```
replace && with `debugConfig?.debugPort`

danarad05 commented:
@amiramw fixed and rebased. some convs - please check response. Thanks

amiramw commented on packages/plugin-dev/src/browser/hosted-plugin-manager-client.ts:
```diff
@@ -306,8 +323,59 @@ export class HostedPluginManagerClient {
+
+        /* if `launchVSCode` is invoked and sessions do not exist - it means that `start` debug was invoked.
+           if `launchVSCode` is invoked and sessions do exist - it means that `restartSessions()` was invoked,
+           which invoked `this.sendRequest('restart', {})`, which restarted `vscode-builtin-js-debug` plugin which is
```
maybe rephrase it with less "which" 😃

danarad05 commented:
@kittaakos are you planning to review this PR? Thanks

danarad05 commented:
@vince-fugnitto  @kittaakos 
cc: @offer8
anyone planning on reviewing this?

danarad05 commented:
@vince-fugnitto @kittaakos @marechal-p
1. Rebased.
2. In continuation of my remark [here](https://github.com/eclipse-theia/theia/pull/9154#issuecomment-844952500) - I would like to complete this PR review ASAP.

Thanks

danarad05 commented:
@vince-fugnitto Are we needing another reviewer approval or is this PR going to be added to this release?

colin-grant-work reviewed (commented):
I've tested this somewhat superficially in base Theia and against our internal debugging setup. Debugging a VSCode plugin in the browser seemed to work fairly well, though I got a notice on session start that a task 'npm' couldn't befound. Testing against our internal setup, I found that the changes to the session `restart` logic seem to have broken scenarios that previously worked - the session is killed but doesn't restart - but I haven't been able to track down exactly where the problem occurs. On that basis, I'd prefer that we wait till after the release to merge this, but then, with a little more time to explore its consequences, I'd be happy to work to get it merged next week.

colin-grant-work commented:
@EstherPerelman, it looks like more conflicts have cropped up in the CHANGELOG. Sorry it's such a pain. Once those are fixed, I have no objection to merging, if others see no problem.

vince-fugnitto commented on packages/debug/src/browser/debug-session.tsx:
```diff
@@ -73,8 +73,12 @@ export class DebugSession implements CompositeTreeElement {
         protected readonly messages: MessageClient,
-        protected readonly fileService: FileService) {
+        protected readonly fileService: FileService,
+        protected readonly debugContributionProvider: ContributionProvider<DebugContribution>
```
Breaking change? Should it be reflected in the `changelog`?
The same would be true for `packages/plugin-ext/src/main/browser/debug/plugin-debug-session-factory.ts` even if it is not likely that someone would extend it.

vince-fugnitto reviewed (approved):
I confirmed that the debugging behavior works fine, and there are no regressions from what I can tell.

vince-fugnitto commented:
@msujew is the pull-request something you're interested in taking a look at, I'm not sure if it impacts the work you've done as part of https://github.com/eclipse-theia/theia/pull/9613.

msujew commented:
@vince-fugnitto I took a quick look at it. I'm expecting some small merge conflicts, but nothing major. Feature-wise this is quite independent of my changes. Thanks for watching out though 👍
